### PR TITLE
Additional activation functions

### PR DESCRIPTION
**Pull request by vloncar**
Adds additional activation functions supported by Keras (ELU, SELU, PReLU, Leaky ReLU, Threshold ReLU, Softplus, Softsign and Hard Sigmoid).

### CODE REVIEW

**Comment by benjaminkreis** [comment]
Hi @vloncar  -- do you already have example Keras models using these activations?

**Comment by vloncar** [comment]
I have tested these functions by changing the activation in [models.py](https://github.com/hls-fpga-machine-learning/keras-training/blob/master/models/models.py) of `keras-training` project. I don't have any other example models.

**Comment by benjaminkreis** [comment]
Ok that works

**Comment by benjaminkreis** [comment]
When you tested, did you run the default build script that includes csynth?

```
csim_design
csynth_design
cosim_design -trace_level all
export_design -format ip_catalog
```

**Comment by vloncar** [comment]
Yes, I have used the default build_prj script that includes those steps.

**Comment by benjaminkreis** [comment]
Which model did you modify?  The `KERAS_3layer` one?

I noticed while testing this PR that the `KERAS_1layer` example is no longer synthesizing, but I found it is also true in the HEAD.  So a recent merged PR must be causing a problem with something unique to `KERAS_1layer` (probably the sigmoid?).  Have you tried sigmoid recently?

**Comment by vloncar** [comment]
I have used the `three_layer_model`. Most of my testing was done before #66 (and later) was merged, however I also quickly tested afterwards and everything was synthesizing properly. I have just tried sigmoid on the project I use for testing the merge layers (concatenate, etc), which is based on sublayer example project, and it synthesized.

**Comment by zhenbinwu** [comment]
@vloncar When you synthesizing the sigmoid, does the HLS try to synthesize the init_sigmoid_table. From my test, it seems HLS start to synthesize this function whenever I use static variable, which takes forever. Surprisingly, it doesn't happen for softmax! And I can't find difference between sigmoid and softmax that would cause this. Thanks!

**Comment by vloncar** [comment]
@zhenbinwu Looking at the logs, I don't see `init_sigmoid_table`. The synthesis taking forever happened to me as well a few times since the introduction of static variables, but not always.

**Comment by zhenbinwu** [comment]
@vloncar So it comes and goes? Any changes between the two cases?

**Comment by vloncar** [comment]
@zhenbinwu I did not investigate at the time, and unfortunately, I don't have logs of failed synthesization attempts. I will try to reproduce the problem and report back, maybe we'll have to find an alternative for the use of static variables.

**Comment by vloncar** [comment]
@zhenbinwu Weirdly enough, if I change the `exp()` to `expf()` in `sigmoid_fcn_float`, or use the `std` namespace (e.g., `std::exp()`, which would be the proper way if we had used `<cmath>` and not `<math.h>`), the synthesis completes. But this makes little sense in relation to the static qualifier for the table variable.

BTW, if tables for activations are static, i.e, there is one shared instance, initialized with the size of `CONFIG_T::table_size` parameter of first activation of that type. In all examples this parameter is the same for all activations, but in general case/future, can it be different? Then the static table would not work as expected. Should we worry about this now?

**Comment by nhanvtran** [comment]
hi @vloncar @zhenbinwu thanks for digging into this.  @vloncar I agree we should have each of the tables be able to be configured differently.  Can we implement a table_size_N for each activation which defaults to the same table size but can be different?  I don't think we save that much by making the table size static right?

**Comment by zhenbinwu** [comment]
@vloncar Interesting and odd! Both sigmoid and softmax called the exp() function with float as input. But only sigmoid has issue and can be fixed by expf().
Static variables are supported by HLS. So this sounds more like a compiler issue then coding to me.

For the multiple activation functions with same instance, from my discussion with @benjaminkreis , I was told the IO to the BRAM are limited to 2. Once pipeline, the IO might be an issue.

**Comment by vloncar** [comment]
@nhanvtran We can make a static table for every unique `config::table_size`. For example, if we have 3 sigmoid activations, two with `table_size = 1024` and one with 512, we would create two tables, one of size 1024 that would be used by two activations, and one with 512 elements. Don't know how complicated would this be given all the limitations of HLS. A simpler alternative would be to create a static table for every activation that needs it, that would save us from reinitialization on multiple invocations of the top level function (which I assume was the original intent of the static tables), at the expense of increased resource usage.

@zhenbinwu I also think this is a HLS compiler issue. Do you think that we can use the `expf` or `std::` workaround as the solution to this problem, or do you think we should investigate further?

**Comment by zhenbinwu** [comment]
@vloncar I think the expf() or std:: workaround should be fine. Don't think we can/have extra time to dig further. Can you make a PR of this? Thanks!

**Comment by vloncar** [comment]
@zhenbinwu Created #78

**Comment by nhanvtran** [comment]
with the merging of the static LUT PR, we'll have to resolve these conflicts one more time.  sorry @vloncar !

File: nnet_utils/nnet_activation.h [hunk 1]
```diff
@@ -325,6 +320,316 @@ void  tanh(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
...
+    for (int ii=0; ii<CONFIG_T::n_in; ii++) {
+        if (CONFIG_T::io_type == io_serial){
+            #pragma HLS PIPELINE
+        }
+        datareg = data[ii];
+        if (datareg > 0) {
+            res[ii] = datareg;
+        } else {
+            index = datareg*CONFIG_T::table_size/8;
+            if (index < 0) index = 0;
```
Comment (benjaminkreis) [inline nnet_utils/nnet_activation.h]: Just tested elu and found that we will always pass this if statement because the indices go from -8 to 0.  Can change this section or the indexing to fix the sign problem.

**Comment by vloncar** [comment]
@benjaminkreis I don't know if it can be incorporated into the existing layer with any syntax similar to the one you posted, but I didn't really explore that option. I have tested it as an additional layer, but keras-to-hls.py should pick up both ways of defining ELU and do the same thing (i.e. there is no additional overhead if activations are defined as layers).

As for the other issue with ELU you noticed, you are right, I will fix that and update the PR when I get back at CERN next week.

File: hls-writer/hls_writer.py [hunk 2]
```diff
@@ -348,6 +370,7 @@ def hls_writer(layer_list, yamlConfig):
             newline += 'typedef {precision} accum_default_t;\n'.format(precision=yamlConfig["DefaultPrecision"])
             newline += 'typedef {precision} weight_default_t;\n'.format(precision=yamlConfig["DefaultPrecision"])
             newline += 'typedef {precision} bias_default_t;\n'.format(precision=yamlConfig["DefaultPrecision"])
+            #newline += 'typedef {precision} alpha_default_t;\n'.format(precision=yamlConfig["DefaultPrecision"])
```
[the remark this answers is not in the document]
Comment (vloncar) [reply hls-writer/hls_writer.py]: I have left this commented out instead of removing it because I wanted to discuss if using separate precision for alpha is worthwhile, but I am not sure if it makes sense, so I will remove it.

[the remark this answers is not in the document]
Comment (benjaminkreis) [reply hls-writer/hls_writer.py]: Ok, understood.  Yeah, I think it's good the way it is.

File: nnet_utils/nnet_activation.h [hunk 3]
```diff
@@ -325,6 +320,316 @@ void  tanh(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
...
+        float in_val = -8.0*ii/float(N_TABLE);
+        // Next, compute lookup table function
+        typename CONFIG_T::table_t real_val = elu_fcn_float(in_val);
+        //std::cout << "Lookup table In Value: " << in_val << " Result: " << real_val << std::endl;
+        table_out[ii] = real_val;
+    }
+}
+
+template<class data_T, class res_T, typename CONFIG_T>
+void  elu(data_T data[CONFIG_T::n_in], res_T res[CONFIG_T::n_in])
```
[the remark this answers is not in the document]
Comment (vloncar) [reply nnet_utils/nnet_activation.h]: I have kept them separated because, like you have observed, elu_alpha has larger latency. If I find the solution to this issue, I will merge them.

**Comment by vloncar** [comment]
I have investigated the hard_sigmoid issue, and strangely enough, the problem was in the way the bounds are checked. `if (datareg < -2.5) ...` doesn't get optimized away, so some weird conversions happen. If I define `data_T left = (data_T) -2.5;` and then do `if (datareg < left)` the latency comes down to 2. Interestingly, if 0 or 1 appear in the if statement, there is no problem, only if we use float values like 2.5. In light of that, I have also made another implementation, like:
```
datareg = slope * data[ii] + shift;
if (datareg > 1) datareg = 1;
else if (datareg < 0) datareg = 0;
res[ii] = datareg;
```
Which has the same performance.

**Comment by vloncar** [comment]
The commit above adds the indexing fixes for `elu` & co, as well as reduces the latency of `hard_sigmoid` and `softsign`. I have used the LUT for `softsign` as suggested, and the change in `hard_sigmoid` is explained above.

I couldn't solve the different latency of `elu` and `elu_alpha`. What I have discovered is that if `alpha=1`, `elu_alpha` has the same latency as `elu`. Analyzing the function (by opening it in analysis perspective in Vivado HLS GUI) shows that the extra multiplication is present if `alpha != 1`, which is the cause for the bigger latency. Therefore I have made one `elu` which can be called with or without alpha (if called without, it is the same as calling it with `alpha=1`). I tried putting alpha into the table initialization, but that blows up latency to 17. On the other hand, in `selu` the `alpha` and `scale` parameters are constants, and can be placed into the LUT, which reduces its latency to 2.

**Comment by benjaminkreis** [comment]
We have merge conflicts here now :(

**Comment by vloncar** [comment]
Rebased. Unfortunately, I made a few mistakes during the rebase (started rebasing twice without completing, git didn't like that), so I had to squash everything into a single commit.

**Comment by benjaminkreis** [comment]
Looks good, @vloncar !

I retested the problematic ones (hard sigmoid, soft sign, elu, alu alpha, and selu), and they all look ok.  Merging!
